Exclude directories from beeing processed

mmattel opened the pull request:
Please see also #15478 (Functionality: exlcude directories from beeing processed) which I needed to close due to a git problem I was not able to solve.

Summary:
1. Enterprise Storage Systems are capable of Snapshots. These Snapshots are directories with a particular name and keep point in time views of the data. Usually, these snapshot directories are present in each directory
2. There is no common naming for these directories and will never be. NetApp uses `.snapshot` and `~snapshot`, EMC eg `.ckpt`, HDS eg `.latest` and `~latest` and so on
3. Viewing and scanning of these directories does not make any sense as these directories are used to restore files from an OS point of view
4. You can also have hardlinks to backups with the same background as described in #15385
5. Scanning a huge datastore containing snapshots does mean that for each snapshot taken ALL files of the datastore are imported into the database again. Example: a datastore containing 200K files, having 15 snapshots means scanning of 3M files where 2.8M are worthless to be imported

There are for sure more resons, but I tried to keep it short.

With this PR, you can define a array of directory names in config.php which are further excluded from processing. It extends the current functionality of excluding `.` and `..`

@DeepDiver1975 @karlitschek @nickvergessen @Aesculapius

icewind1991 committed:
unit test for etag propagation across shared storage boundary

icewind1991 committed:
Make the change propagator an emitter

icewind1991 committed:
Allow getting *all* share entries owned by a user

icewind1991 committed:
propagate etags for all user of a share

icewind1991 committed:
dont use our now non existing hook

icewind1991 committed:
adjust test

mmattel committed:
Improve error messge text for app upgrade try (#15375)

icewind1991 committed:
remove duplicates from rebase

Commit:
Fix version rollback to keep fileid

rullzer committed:
Only show SSL table header if there are certificates

Commit:
skip user if we don't have a public key

nickvergessen committed:
Correctly generate the feedback URL for remote share

The trailing slash was added in c78e3c4a7fa1d2f474ab58551e67a50e093f6ed8
to correctly generate the encryption keys

icewind1991 committed:
more propagation tests

icewind1991 committed:
Make sure the owner mountpoints are setup

icewind1991 committed:
propagate changes up reshares

oparoz committed:
Show a text preview instead of a bitmap preview of text
+ fix the preview height

nickvergessen committed:
Verify that the encryption module exists before setting it

nickvergessen committed:
Add occ commands to manage the encryption modules

oparoz committed:
Cache the token value

oparoz committed:
Use webDAV to have access to HTTP byte ranges

nickvergessen committed:
Add a test that the default module is returned before we fall back

nickvergessen committed:
Add test for setDefaultEncryptionModule

nickvergessen committed:
Deduplicate module mock

icewind1991 committed:
add rename hook for propagation

oparoz committed:
An idea to suggest that this is just a sample of the full text

nickvergessen committed:
Make getDefaultModuleId public and get module protected

nickvergessen committed:
Make methods explicit public

icewind1991 committed:
minor fixes

th3fallen committed:
fixing return values and adding tests

icewind1991 committed:
fix propagation when renaming a directly reshared folder

icewind1991 committed:
fix merge

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/aa51444c-1d42-4568-bd23-0c407706906e) was created.

oparoz committed:
Use authorization headers for better compatibility
+ add ellipsis
+ fix OC.generateUrl calls

th3fallen committed:
destupify tests

karlitschek commented:
@LukasReschke @DeepDiver1975 @icewind1991

nickvergessen commented on lib/private/files/filesystem.php:
```diff
@@ -584,6 +585,14 @@ static public function isIgnoredDir($dir) {
+				}
+			}
+		}
 		return false;
```
Why don't you use `return in_array($dir, $excluded);` like we do for the files option?

nickvergessen committed:
Do not restrict permissions for the original owner

blizzz committed:
LDAP Wizard: have always-increasing 'nth Server' number per full page load, fixes #15766

blizzz committed:
save configs when requesting a config ID. They are empty, but avoid configID collisioning when creating many new configs in the wizard without saving anything directly.

icewind1991 committed:
triger propagation for webdav uploads

use post hooks for share etag propagator

Commit:
fix encryption header error

When moving back the pointer to position 0 (using stream_seek), the pointer on the encrypted stream will be moved to the position immediately after the header. Reading the header again (invoked by stream_read) will cause an error, writing the header again (invoked by stream_write) will corrupt the file. Reading/writing the header should therefore happen when opening the file rather than upon read or write. Note that a side-effect of this PR is that empty files will still get an encryption header; I think that is OK, but it is different from how it was originally implemented.

Commit:
Update encryption.php

icewind1991 committed:
fix test

Commit:
add two tests

testRewind tests reading and writing after rewind on an encrypted stream; testWriteWriteRead tests r+ mode

Commit:
fixed name

Commit:
disable r+ test

Commit:
Update encryption.php

Commit:
Added unit test for reshare by share owner

Commit:
Update encryption.php

Commit:
enable testRewind

Commit:
enable testWriteWriteRead

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/403026cb-240b-45c5-870a-86d181e9d628) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/403026cb-240b-45c5-870a-86d181e9d628) completed: **No new issues**

LukasReschke commented:
I'm still not absolutely sure whether this is the best way to implement this, for one we need to ensure that all code paths call `isIgnoredDir` so that you can't upload files there as such invisible files would be very confusing for users. cc @icewind1991

LukasReschke commented:
(also requires unit tests)

icewind1991 committed:
use the permissions mask cache wrapper instead of the read only cache

icewind1991 committed:
resolve reshares in public webdav

th3fallen committed:
remove status's and adjust js

mmattel commented:
It is impotant that we talk about directories and not files. What means that if a directory is not "visible" via owncloud, you can not write into.

In the first step, it is a matter of scanning. What is not scanned is not shown and further not processed.

But you are right about the ``creation of a directory via owncloud`` with an excluded directory name, that should be covered.

In filesystem.php, we have the function:

	static public function mkdir($path) {
		return self::$defaultInstance->mkdir($path);
	}

What about if we do:

	static public function mkdir($path) {
		if (self::isIgnoredDir(basename($path))) {
			return false;
		}
		return self::$defaultInstance->mkdir($path);
	}

I have tested this and this works. I get a owncloud error in case a excluded directory is chosen:
![image](https://cloud.githubusercontent.com/assets/3321281/7278534/794a4b9c-e916-11e4-8a54-29899ca46d1e.png)

If I comment out the code of ``mkdir`` I added and try to create a directory called ``~snapshot`` into a SMB share which is on a storage having snapshot directory names like ``~snapshot`` I get: 
![image](https://cloud.githubusercontent.com/assets/3321281/7278883/79cf475a-e918-11e4-9523-c90717c06b61.png)
This means, that the code above would perfectly work. Because anyway, the directory can not be created via the owncloud - it returns an error.
Please let me know your thoughts.

But it also would be great to cover that in an earler stage eg when the path is entered: https://github.com/owncloud/core/blob/master/apps/files/js/files.js#L97
Here I do not know how to do it, I would need some help.

th3fallen committed:
fix messages from settings crontroller

icewind1991 committed:
update icewind/smb to 1.0.1

icewind1991 committed:
remove unneeded readonlycache

nickvergessen commented:
@mmattel but we have several ways of doing that, webdav, provisioning api, web ui, ...
and that is what lukas means, it's hard to make sure that all use this way now all of a sudden.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/e53da423-626f-42a9-afac-54997454c780) was created.

mmattel commented:
OK, I got that. Step by step.

I added a check at the webgui for excluded directories and that works. If you try to create a directory via the webgui with a name in the excluded list, you get a message presented.

Where can I look next to adopt the code for implementing this?

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/e53da423-626f-42a9-afac-54997454c780) completed: **2 new issues**

nickvergessen committed:
Check if the user exists before trying to delete him

nickvergessen committed:
Add unit tests for occ user:delete

nickvergessen committed:
Unify the output of the user commands and use DI

nickvergessen committed:
Add tests for occ user:lastseen

nickvergessen committed:
Fix file names

Commit:
Delete temp files after testing encryption stream wrapper

icewind1991 committed:
fix subfolder reshares over webdav

LukasReschke committed:
Fix collision on temporary files + adjust permissions

This changeset hardens the temporary file and directory creation to address multiple problems that may lead to exposure of files to other users, data loss or other unexpected behaviour that is impossible to debug.

**[CWE-668: Exposure of Resource to Wrong Sphere](https://cwe.mitre.org/data/definitions/668.html)**
The temporary file and folder handling as implemented in ownCloud is performed using a MD5 hash over `time()` concatenated with `rand()`. This is insufficiently and leads to the following security problems:
The generated filename could already be used by another user. It is not verified whether the file is already used and thus temporary files might be used for another user as well resulting in all possible stuff such as "user has file of other user".

Effectively this leaves us with:

1. A timestamp based on seconds (no entropy at all)
2. `rand()` which returns usually a number between 0 and 2,147,483,647

Considering the birthday paradox and that we use this method quite often (especially when handling external storage) this is quite error prone and needs to get addressed.

This behaviour has been fixed by using `tempnam` instead for single temporary files. For creating temporary directories an additional postfix will be appended, the solution is for directories still not absolutely bulletproof but the best I can think about at the moment. Improvement suggestions are welcome.

**[CWE-378: Creation of Temporary File With Insecure Permissions](https://cwe.mitre.org/data/definitions/378.html)**

Files were created using `touch()` which defaults to a permission of 0644. Thus other users on the machine may read potentially sensitive information as `/tmp/` is world-readable. However, ownCloud always encourages users to use a dedicated machine to run the ownCloud instance and thus this is no a high severe issue. Permissions have been adjusted to 0600.

**[CWE-379: Creation of Temporary File in Directory with Incorrect Permissions](https://cwe.mitre.org/data/definitions/379.html)**

Files were created using `mkdir()` which defaults to a permission of 0777. Thus other users on the machine may read potentially sensitive information as `/tmp/` is world-readable. However, ownCloud always encourages users to use a dedicated machine to run the ownCloud instance and thus this is no a high severe issue. Permissions have been adjusted to 0700.Please enter the commit message for your changes.

MorrisJobke committed:
Show hint if there is no delete permission

* add class .no-permission which shows the default delete icon
* fixes #15172
* add unit test for no permission and for delete permission

DeepDiver1975 committed:
Merge pull request #15617 from mmattel/getenv_system_check_and_warning

Checks if getenv returns proper system environment variable results

DeepDiver1975 committed:
Merge pull request #15831 from owncloud/enc2-unittestcleanuptmpfile

Delete temp files after testing encryption stream wrapper

DeepDiver1975 committed:
Merge pull request #15828 from owncloud/smb-1.0.1

update icewind/smb to 1.0.1

Commit:
Merge pull request #15814 from owncloud/public-reshare-webdav

Fix webdav access for public reshare

Commit:
update share keys if a file is moved to a shared folder

DeepDiver1975 committed:
Ignore test folders when checking the code for compliance

Commit:
update share keys if file gets copied

nickvergessen committed:
Add unit tests for the commands

oparoz committed:
Remove overflow+watermark

icewind1991 committed:
cleanup hooks

icewind1991 committed:
use microtime to prevent race conditions

MorrisJobke committed:
Add missing loading indicator for enabled apps

* happened when an app gets activated, because the
  new appended HTML doesn't contain the loading
  image
* fixes #15806

blizzz committed:
Merge pull request #15826 from owncloud/issue-15804-occ-user-delete-exception

Issue 15804 occ user delete exception

LukasReschke committed:
Merge pull request #15846 from owncloud/fix-missing-loading-indicator

Add missing loading indicator for enabled apps

Commit:
[tx-robot] updated from transifex

Commit:
only update share keys if the file was encrypted

jancborchardt committed:
Merge pull request #15835 from owncloud/delete-permission-icon

Show hint if there is no delete permission

Commit:
Tear down FS after running trash expire

Commit:
fall back to the ownCloud default encryption module and aes128 if we read a encrypted file without a header

nickvergessen committed:
Use public interfaces for type hinting

nickvergessen committed:
Deduplicate data provider and fix method visibility

LukasReschke committed:
Merge pull request #15855 from owncloud/teardownaftercommand

Tear down FS after running trash expire

nickvergessen committed:
Add test "operation on keys failed"

nickvergessen committed:
Merge pull request #15838 from owncloud/code-checker-to-ignore-tests

Ignore test folders when checking the code for compliance

nickvergessen committed:
Merge pull request #15839 from owncloud/enc_fix_moving_shared_files

[encryption] fix moving files to a shared folder

th3fallen committed:
change error codes to 400

Commit:
fix unit tests

Commit:
Merge pull request #15774 from owncloud/jknockaert-patch-1

fix encryption header error

icewind1991 committed:
safer teardown

LukasReschke committed:
Merge pull request #15683 from owncloud/block-legacy-clients

Block old legacy clients

Commit:
[tx-robot] updated from transifex

LukasReschke committed:
Merge pull request #15834 from owncloud/make-temporary-file-really-unique

Fix collision on temporary files + adjust permissions

Commit:
[tx-robot] updated from transifex

jancborchardt committed:
fix layout for multiline buttons which are <a class='button'>

Commit:
[tx-robot] updated from transifex

nickvergessen committed:
Merge pull request #15875 from owncloud/multiline-buttons

fix layout for multiline buttons which are <a class='button'>

nickvergessen committed:
Use interface

nickvergessen committed:
Fix new tests

nickvergessen committed:
Fix scrutinizer complain

nickvergessen committed:
Fix module registration

nickvergessen committed:
Undefined index: default

MorrisJobke committed:
Merge pull request #15681 from owncloud/issue/15667-occ-encryption

OCC commands to manage encryption

LukasReschke committed:
Fix type annotation

Obviously should be an int

DeepDiver1975 committed:
Merge pull request #15411 from mmattel/fix_for_15375_better_message_text

Improve error messge text for app upgrade try (#15375)

MorrisJobke committed:
Merge pull request #15882 from owncloud/fix-type-annotation

Fix type annotation

DeepDiver1975 committed:
fixes #15848

nickvergessen committed:
Can't use function return value in write context

DeepDiver1975 committed:
Merge pull request #15860 from owncloud/enc_fallback_old_encryption

[encryption] handle encrypted files correctly which where encrypted with a old version of ownCloud (<=oc6)

nickvergessen committed:
Fix several type(hint) errors in private/helper.php

DeepDiver1975 committed:
Merge pull request #15886 from owncloud/fix-15848-master

Adjust isLocal() on encryption wrapper

DeepDiver1975 committed:
Merge pull request #15887 from owncloud/issue-15885-empty-getenv-path

Can't use function return value in write context

icewind1991 committed:
dont pass floats as timestamp to the changepropagator

LukasReschke committed:
Rephrase "TL;DR"

"TL;DR" seems not to be understood or respected by everyone. Maybe we should do a try with "Important". Open for alternative suggestions.

nickvergessen committed:
Allow setting protected properties

nickvergessen committed:
Allow setting protected properties

nickvergessen committed:
Simplify the test

DeepDiver1975 committed:
Merge pull request #15890 from owncloud/fix-helper-docs

Fix several type(hint) errors in private/helper.php

LukasReschke committed:
Simplify even more

DeepDiver1975 committed:
handle case where the given language is unknown - refs #15794

Commit:
Fix file name validation in New menu

DeepDiver1975 committed:
fix typo

nickvergessen committed:
check the called URL

Commit:
Add check for "replace" command

LukasReschke committed:
Merge pull request #15891 from owncloud/LukasReschke-patch-1

Rephrase "TL;DR"

nickvergessen committed:
Add tests for the remote sharing url

nickvergessen committed:
Correctly remove the protocol before prepeding it

MorrisJobke committed:
Merge pull request #15893 from owncloud/files-fixnewfilevalidation

Fix file name validation in New menu

Commit:
[tx-robot] updated from transifex

nickvergessen committed:
make scrutinizer happy

nickvergessen committed:
fix several issues with doc blocks on share.php

nickvergessen committed:
Fix scrutinizer complains and return type doc

nickvergessen committed:
Fix return type of the getRootFolder() method

DeepDiver1975 committed:
Merge pull request #15903 from owncloud/issue-15899-fix-getfoortfolder-doc

Fix return type of the getRootFolder() method

rullzer committed:
Password set via OCS API should not be double escaped

MorrisJobke committed:
Merge pull request #15901 from owncloud/fix-share-docs

fix several issues with doc blocks on share.php

DeepDiver1975 committed:
Merge pull request #15896 from owncloud/autotest-external-replacecommand

Add check for "replace" command in autotest-external

DeepDiver1975 committed:
Merge pull request #15773 from owncloud/fix-15766

LDAP Wizard: increasing server numbers and avoid configID collisions, fixes #15766

MorrisJobke committed:
Merge pull request #14764 from owncloud/shared-etag-propagate

Propagate etags across shared storages

MorrisJobke committed:
Replace `replace` with `sed` statement

DeepDiver1975 committed:
Merge pull request #15892 from owncloud/fix-failed-require-in-l10n-createjs

handle case where the given language is unknown - refs #15794

Commit:
make sure that we create a header for the encrypted private key

rullzer committed:
Added unit test

nickvergessen committed:
Correctly send Federate-Cloud Share Feedback against the remote share ID

LukasReschke committed:
Fix routes

Fixes public sharing as reported on https://github.com/owncloud/core/issues/15913

nickvergessen committed:
Add tests for the correct share id on the call aswell

icewind1991 committed:
extends memcache with add, inc and dec

DeepDiver1975 committed:
Merge pull request #15915 from owncloud/fix-sharing

Fix routes

LukasReschke committed:
Filter potential dangerous filenames for avatars

We don't want to have users misusing this API resulting in a potential file disclosure of "avatar.(jpg|png)" files.

Commit:
skip update of encryption keys if file is not encrypted

Commit:
also create encryption keys for empty files

Commit:
method shouldn't be static

cmeh committed:
Update wizardTabElementary.js

Corrected "More then …" to "More than …".

JostBaron committed:
Fix error message when wrong user on console

Owncloud complains if the user runs a shell script with a user other than the user that owns `config/config.php`. The error message does not reflect this, and instead speaks about the web server user. Related issue: #15920 .

Commit:
[tx-robot] updated from transifex

nickvergessen committed:
Merge pull request #15918 from owncloud/harden-avatar

Filter potential dangerous filenames for avatars

DeepDiver1975 committed:
Merge pull request #15921 from cmeh/patch-4

Spelling in wizardTabElementary.js: Correct "More then …" to "More than …"

DeepDiver1975 committed:
Merge pull request #15922 from JostBaron/patch-1

Fix error message when wrong user on console

LukasReschke committed:
Merge pull request #15706 from owncloud/fix-15705-master

move creation of shared key to the login handler

schiessle committed:
Merge pull request #15911 from owncloud/enc_create_header_for_new_key

make sure that we create a header for the encrypted private key

MorrisJobke committed:
Merge pull request #15906 from rullzer/fix_15777

Password set via OCS API should not be double escaped

MorrisJobke committed:
Merge pull request #15596 from owncloud/issue/15589

Correctly generate the feedback URL for remote share

Commit:
default to an empty string for database password if ommited

fix null passwords

remove password trimming

Commit:
Merge pull request #15908 from owncloud/fix-replace-with-sed

Replace `replace` with `sed` statement

nickvergessen committed:
Add a test to share a subfolder of a folder shared with the owner by group

nickvergessen committed:
Correctly select file cache values also when checking group shares

nickvergessen committed:
Ignore "parent" shares when the sharee is the owner of the reshare-source

icewind1991 committed:
add compare and swap to memcache

Commit:
check recovery setting for the right user

icewind1991 committed:
add memcache based shared/exclusive locking

jancborchardt committed:
Revert "fix layout for multiline buttons which are <a class='button'>"

jancborchardt committed:
add title to multiselect list items so longer group names are visible, fix #15819

DeepDiver1975 committed:
Merge pull request #15926 from owncloud/install-empty-pw

Make it possible to pass empty database passwords to install command

Commit:
implement json serialize for l10n string

Commit:
add aliases to pascal case constructor paramters to make it possible to auto assemble controllers

Commit:
make command optional

th3fallen committed:
Merge pull request #15919 from owncloud/enc_handle_empty_files

Encryption improve handling of empty and unencrypted files

nickvergessen committed:
Merge pull request #15942 from owncloud/revert-15875-multiline-buttons

Revert "fix layout for multiline buttons which are <a class='button'>"

Commit:
[tx-robot] updated from transifex

MorrisJobke committed:
Merge pull request #15947 from owncloud/add-title-to-multiselect

add title to multiselect list items so longer group names are visible

DeepDiver1975 committed:
Merge pull request #15946 from owncloud/auto-controller-resolve

Make controller automatically assemblable

MorrisJobke committed:
Merge pull request #15951 from owncloud/fix-install-again

Make command optional

DeepDiver1975 committed:
Moving encryption_dummy to the apps repo - fixes #15950

DeepDiver1975 committed:
Merge pull request #15949 from owncloud/l10n-string-json

Implement json serialize for l10n string

MorrisJobke committed:
Use internally \OCP\ILogger instead of \OC\Log

* this is the preparation for some upcoming logger related changes
* also fixes an issue in the public interface where we request
  an internal class as parameter

Commit:
fix #15962

DeepDiver1975 committed:
Enable code coverage on scrutinizer

icewind1991 committed:
use the preSetup hook to ensure the encryption wrapper is applied correctly

icewind1991 committed:
use trait for cas polyfill for xcache

icewind1991 committed:
add tests for castrait

DeepDiver1975 committed:
Merge pull request #15968 from owncloud/enc-wrapper-presetup

use the preSetup hook to ensure the encryption wrapper is applied correctly

icewind1991 committed:
allow getting the path from the lockedexception

DeepDiver1975 committed:
Merge pull request #15964 from owncloud/fix-l10n-string-warn

Fix #15962

Commit:
Use moveFromStorage instead of streamCopy when restoring version

DeepDiver1975 committed:
Merge pull request #15972 from owncloud/enable-scrutinizer-code-coverage

Enable code coverage on scrutinizer

DeepDiver1975 committed:
Merge pull request #15955 from owncloud/move-dummy-encryption

Moving encryption_dummy to the apps repo

jancborchardt committed:
add admin documentation links for Sharing, Encryption and Email server

jancborchardt committed:
fix layout for documentation link hints

MorrisJobke committed:
Merge pull request #15993 from owncloud/admin-documentation

add admin documentation links for Sharing, Encryption and Email server

Commit:
[tx-robot] updated from transifex

DeepDiver1975 committed:
Merge pull request #15937 from owncloud/file-locking

Add memcache based shared/exclusive locking

jancborchardt committed:
also add title to multiselect create new element, fix #15954

jancborchardt committed:
force minimum multiselect width to properly show longer groups

jancborchardt committed:
remove 'More apps' link from apps mgmt now that we have support for experimental apps

Commit:
[tx-robot] updated from transifex

jancborchardt committed:
fix search results from other places, account for margin-bottom in files list

LukasReschke committed:
Merge pull request #16005 from owncloud/remove-more-apps

remove 'More apps' link from apps mgmt now that we have support for experimental apps

LukasReschke committed:
Add proper versioning for doc URL

As per 8.1 we have docs for minor versions as well so we need to link to `8.1` here instead to `8.0`.

Fixes https://github.com/owncloud/core/issues/16002

MorrisJobke committed:
Merge pull request #16015 from owncloud/add-proper-versioning

Add proper versioning for doc URL

MorrisJobke committed:
Merge pull request #16001 from owncloud/add-multiselect-new-title

also add title to multiselect create new element, force multiselect wider width, fix #15954

Commit:
[tx-robot] updated from transifex

MorrisJobke committed:
Merge pull request #16009 from owncloud/fix-searchresults

fix search results from other places, account for margin-bottom in files list

MorrisJobke committed:
Merge pull request #15486 from owncloud/versions-rollbackkeepfileid

Keep fileid when restoring version, using stream copy

LukasReschke committed:
Try to set required ini values in base.php

Fixes https://github.com/owncloud/core/issues/16006

Commit:
8.1 alpha 3

Commit:
[tx-robot] updated from transifex

DeepDiver1975 committed:
Merge pull request #15958 from owncloud/usage-of-public-log-interface

Use internally \OCP\ILogger instead of \OC\Log

DeepDiver1975 committed:
Merge pull request #16020 from owncloud/set-charset-via-ini-set

Try to set required ini values in base.php

DeepDiver1975 committed:
Merge pull request #15584 from owncloud/enc_fix_upload_shared_folder

skip user if we don't have a public key

Commit:
some small ux improvements

DeepDiver1975 committed:
no link if there is no detailpage property set - fixes #15953

DeepDiver1975 committed:
fix l10n strings

DeepDiver1975 committed:
Merge pull request #15752 from owncloud/feature/fix-encryption-return-values

[enc2] fixing return values and adding tests

LukasReschke committed:
Merge pull request #15652 from oparoz/better-text-previews

Show a text preview instead of a bitmap preview of text

MorrisJobke committed:
Fix DeleteOrphanedSharesJob

* remove unneeded use statements
* the the interval properly in the constructor
* set the log level to debug (from info) - fixes #16036

LukasReschke committed:
Remove hard-dependency on disabled output_buffering

This removes the hard-dependency on output buffering as requested at https://github.com/owncloud/core/issues/16013 since a lot of distributions such as Debian and Ubuntu decided to use `4096` instead of the PHP recommended and documented default value of `off`.

However, we still should encourage disabling this setting for improved performance and reliability thus the setting switches in `.user.ini` and `.htaccess` are remaining there. It is very likely that we in other cases also should disable the output buffering but aren't doing it everywhere and thus causing memory problems.

Fixes https://github.com/owncloud/core/issues/16013

LukasReschke committed:
Remove PHP 5.4 warning in checkSetup

This is catched in index.php as older PHP versions will never execute the code path until there due to 5.4 syntax changes.

nickvergessen committed:
Fix undefined variable $tagId

nickvergessen committed:
Merge pull request #15772 from owncloud/issue-15771-dont-restrict-permissions-for-share-owner

Do not restrict permissions for the original owner

Commit:
Combine autotest-hhvm.sh with autotest.sh

nickvergessen committed:
Add an occ command to get the status of encryption

LukasReschke committed:
Check `mbstring.func_overload` only if the mb module is installed.

Fixes https://github.com/owncloud/core/issues/14670

LukasReschke committed:
Merge pull request #16043 from owncloud/activity-288-log-entry-when-no-favorite

Fix undefined variable $tagId

LukasReschke committed:
Merge pull request #16040 from owncloud/fix-orphaned-shares-job

Fix DeleteOrphanedSharesJob

DeepDiver1975 committed:
Merge pull request #16044 from owncloud/issue-15977-occ-encryption-status

Add an occ command to get the status of encryption

DeepDiver1975 committed:
Merge pull request #16042 from owncloud/fix-output-buffering

Remove hard-dependency on disabled output_buffering

schiessle committed:
Merge pull request #16026 from owncloud/enc_ux_improvements

[encryption] some small ux improvements

LukasReschke committed:
Fix PHPDoc for the ext. share controller

LukasReschke committed:
Merge pull request #16048 from owncloud/fix-mb-string

Check `mbstring.func_overload` only if the mb module is installed.

jancborchardt committed:
Merge pull request #15544 from rullzer/fix_15263

[WIP] Only show SSL table header if there are certificates

Commit:
[tx-robot] updated from transifex

nickvergessen committed:
Fix the "app-id" parameter description of app:check-code

jancborchardt committed:
Merge pull request #16031 from owncloud/fix-15953-master

No clickable app for local apps

schiessle committed:
Merge pull request #15938 from owncloud/enc_update_recovery_share_key

[encryption] check recovery key setting for the correct user

nickvergessen committed:
Merge pull request #16052 from owncloud/fix-phpdoc-sharecontroller

Fix PHPDoc for the ext. share controller

MorrisJobke committed:
Merge pull request #16058 from owncloud/fix-app-check-code-parameter-description

Fix the "app-id" parameter description of app:check-code

DeepDiver1975 committed:
Merge pull request #16045 from owncloud/reduce-autotest-hhvm

Combine autotest-hhvm.sh with autotest.sh

blizzz committed:
not only send them back but also apply defaults to a new configuration. Fixes #15933

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/ccb2ba9d-6e82-4d88-8505-7429640bf4e5) was created.

mmattel commented:
Changes to cover webdav, webgui and client access

Following test have been made creating either a file or directory having once a allowed and once a excluded name.

The functions shown in the table are except one all triggered  and covered by the access methods used in lib/private/files/view.php

For WebDav, CyberDuck was used on a W7 client.
The IOS client run the latest available version (3.4.1)
All tests made passed well.
A unit test is open to be made.

		| webdav	| webgui	| client IOS
------------ |------------ |------------ | -------------
directory create | mkdir	| mkdir		| mkdir
directory rename | rename	| rename	| rename
directory delete | rmdir	| rmdir		| rmdir
		 | 		| 		| 
file create	 | put(*)	| touch		| n.a.
file rename	 | rename	| rename	| rename
directory delete | unlink	| unlink	| unlink


n.a. ... function not available
(*) put ... this function is not handled in lib/private/files/view.php  but in lib/private/connector/sabre/directory.php which points to /sabre/file.php

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/e4c8af19-06f3-4b1f-bbed-d7e1fedaceb9) was created.

mmattel commented:
git drives me nuts, it was just a rebase and now I have all those additional commits in...
will try to fix that.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/54062934-eb13-4d21-8af5-2ec282afac2c) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/9fed9b63-de30-49c5-aaf1-1a313b36d84e) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/0c2fb639-0e3f-45e5-aa3c-5d6a5d24baf8) was created.

mmattel commented:
Closing due to a rebase problem will reopen with a new PR.
Sorry for any inconvenience caused
